### PR TITLE
Add a new section named development in issue view sidebar to interact with branch/pr

### LINKED ISSUE #20226: Create a branch from an issue, with pull/merge request
### Feature Description

There is a workflow that uses issues to define tasks to do and then work branches to do the issue's work in. When the work is done, the branch will be merged into master/main. It saves a lot of manual work if the Git tool already creates a work branch from an issue and also a merge/pull request for it with the corresponding "Closes issue ..." message.

GitLab does this and I'm using it a lot. GitHub also does this but that's not (yet?) reflected in Gitea's [feature comparison](https://docs.gitea.io/en-us/comparison/#issue-tracker).

It has been [asked before](https://github.com/go-gitea/gitea/issues/19827) but was closed. Not sure why because it was neither implemented nor rejected and can't be reopened for some reason. Here's a second try.

### GitHub

grafik

grafik

### GitLab

grafik

grafik
(Clicking the button, not its dropdown menu, would immediately create the branch and MR)

### PR DESCRIPTION
**Pull request by lunny**
Partially Resolve #20226 

This PR add a new section on the sidebar of issue named `Development`. It supports creating one or multiple branches and pull requests from the sidebar.

1. No related branch or PR
image

2. Create new branch from the issue
image

3. Related branch created
image

4. Prefill title and content when creating pull request from the issue development link
图片

5. Realted PR created
image

6. Related PR merged
image

7. Mayge fixed near the issue title
image

- [x] Delete dev_link records from the database when deleting repos/issues
- [x] Delete dev_link records from the database when deleting pull requests/branch
- [x] Allow creating multiple branches/pull requests from UI
- [x] Bring issue information into the pull request when creating the pull request.
- [x] Adjust the toast z-index to 9999 so that it will display in front of the modal.

### CODE REVIEW

File: routers/web/repo/issue.go [hunk 1]
```diff
@@ -2078,6 +2078,42 @@ func ViewIssue(ctx *context.Context) {
...
+		devLinks, err := issue_service.FindIssueDevLinksByIssue(ctx, issue)
+		if err != nil {
+			ctx.ServerError("FindIssueDevLinksByIssue", err)
+			return
+		}
+		ctx.Data["DevLinks"] = devLinks
+		for _, link := range devLinks {
+			if link.LinkType == issues_model.IssueDevLinkTypePullRequest {
+				if !(link.PullRequest.Issue.IsClosed && !link.PullRequest.HasMerged) {
+					ctx.Data["MaybeFixed"] = link.PullRequest
```
Comment (Zettat123) [inline routers/web/repo/issue.go]: Will it be conflict with cross references? In some cases, the title may say "May be fixed by #1" and the comment list may say "User referenced a pull request that will close this issue (#2)".

Comment (lunny) [reply routers/web/repo/issue.go]: In future, I will add a new record in this table when there is a xref.

**Review by Zettat123 — APPROVED** [review]

File: templates/repo/issue/view_title.tmpl [hunk 2]
```diff
@@ -123,6 +123,10 @@
 					·
 					{{ctx.Locale.TrN .Issue.NumComments "repo.issues.num_comments_1" "repo.issues.num_comments" .Issue.NumComments}}
 				</span>
+				{{if .MaybeFixed}}
+					{{$fixedStr := printf `<a href="%s">#%d</a>` .MaybeFixed.Issue.Link .MaybeFixed.Index}}
+					· <span>{{ctx.Locale.Tr "repo.issues.maybefixed" ($fixedStr|SafeHTML)}}</span>
```
Comment (wxiaoguang) [inline templates/repo/issue/view_title.tmpl]: There should be no new SafeHTML in code anymore.

Comment (lunny) [reply templates/repo/issue/view_title.tmpl]: 0e0527456edf836b1e77b71f4ac1f3412b6ab08a

**Comment by yp05327** [comment]
Delete repo should also delete these links?
There's no changes in
image

**Comment by lunny** [comment]
> Delete repo should also delete these links? There's no changes in image

This deletion has been added in issue deletion. see updated in `models/issues/issue_update.go`

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1628,6 +1628,15 @@ issues.label.filter_sort.alphabetically = Alphabetically
...
 issues.label.filter_sort.reverse_by_size = Largest size
+issues.development = Development
+issues.maybefixed = May be fixed by %s
+issues.create_branch_from_issue_success = Create branch %s from issue successfully
+issues.base_branch = Base Branch
+issues.pr.completed = Completed
+issues.pr.conflicted = Merge conflicts
+issues.pr.not_exist_issue = Reference issue does not exist.
+issues.branch.latest = Latest commit %s
+issues.link.created = Created %s
```
Comment (yp05327) [inline options/locale/locale_en-US.ini]: Only one word, no need to add `%s`?

File: routers/web/repo/issue.go [hunk 4]
```diff
@@ -2078,6 +2078,42 @@ func ViewIssue(ctx *context.Context) {
...
+
+		devLinks, err := issue_service.FindIssueDevLinksByIssue(ctx, issue)
+		if err != nil {
+			ctx.ServerError("FindIssueDevLinksByIssue", err)
+			return
+		}
+		ctx.Data["DevLinks"] = devLinks
+		for _, link := range devLinks {
+			if link.LinkType == issues_model.IssueDevLinkTypePullRequest {
+				if !(link.PullRequest.Issue.IsClosed && !link.PullRequest.HasMerged) {
```
Comment (yp05327) [inline routers/web/repo/issue.go]: This can be merged

File: options/locale/locale_en-US.ini (hunk 3 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: Because different languages have different sequences, some would like to put time later but some put front.

File: routers/web/repo/issue.go (hunk 4 above)
Comment (lunny) [reply routers/web/repo/issue.go]: How?

File: routers/web/repo/issue_dev.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+func CreateBranchFromIssue(ctx *context.Context) {
+	issue := GetActionIssue(ctx)
+	if ctx.Written() {
+		return
+	}
+
+	if issue.IsPull {
+		ctx.Flash.Error(ctx.Tr("repo.issues.create_branch_from_issue_error_is_pull"))
```
Comment (yp05327) [inline routers/web/repo/issue_dev.go:34]: I can not find this in `options/locale/locale_en-US.ini`

File: routers/web/repo/issue_dev.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	if form.RepoID > 0 {
+		var err error
+		repo, err = repo_model.GetRepositoryByID(ctx, form.RepoID)
+		if err != nil {
+			ctx.ServerError("GetRepositoryByID", err)
+			return
+		}
+		gitRepo, err = gitrepo.OpenRepository(ctx, repo)
+		if err != nil {
+			ctx.ServerError("GetRepositoryByID", err)
```
Comment (yp05327) [inline routers/web/repo/issue_dev.go]:
```suggestion
			ctx.ServerError("OpenRepository", err)
```

File: routers/web/repo/issue_dev.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+			return
+		}
+		defer gitRepo.Close()
+	}
+
+	perm, err := access_model.GetUserRepoPermission(ctx, repo, ctx.Doer)
+	if err != nil {
+		ctx.ServerError("GetRepositoryByID", err)
```
Comment (yp05327) [inline routers/web/repo/issue_dev.go]:
```suggestion
		ctx.ServerError("GetUserRepoPermission", err)
```

File: routers/web/repo/issue_dev.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+		return
+	}
+
+	canCreateBranch := perm.CanWrite(unit_model.TypeCode) && repo.CanCreateBranch()
+	if !canCreateBranch {
+		ctx.NotFound("CreateBranch", nil)
```
Comment (yp05327) [inline routers/web/repo/issue_dev.go]: 404 or 403?

File: routers/web/repo/issue_dev.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+		return
+	}
+
+	if ctx.HasError() {
+		ctx.JSONError(ctx.GetErrMsg())
+		return
+	}
```
Comment (yp05327) [inline routers/web/repo/issue_dev.go]: What is this for?

File: routers/web/repo/issue_dev.go (hunk 5 above)
Comment (lunny) [reply routers/web/repo/issue_dev.go:34]: 7fd210b73374b8f3c68af1f2c39a4d98304ba924

File: routers/web/repo/issue_dev.go (hunk 9 above)
Comment (lunny) [reply routers/web/repo/issue_dev.go]: ref 30d4010983c372d05d3e53c23fe5539022835548

File: routers/web/repo/issue_dev.go (hunk 8 above)
Comment (lunny) [reply routers/web/repo/issue_dev.go]: It should be 403. 30d4010983c372d05d3e53c23fe5539022835548

File: routers/web/repo/issue.go (hunk 4 above)
Comment (yp05327) [reply routers/web/repo/issue.go]: if link.LinkType == issues_model.IssueDevLinkTypePullRequest && !link.PullRequest.Issue.IsClosed && !link.PullRequest.HasMerged

Do you mean `not closed and not merged PRs`? The `!` seems wrong. The first one is outside of the `(xxxx)`

Comment (lunny) [reply routers/web/repo/issue.go]: No. This means opened pull request or closed but merged pull request. Only closed and not merged pull request should not be considered as a maybe-fixed pull request.

Comment (lunny) [reply routers/web/repo/issue.go]: 66681c33ffd42f636528a40a0b0c3fcd1a37aba2

Comment (yp05327) [reply routers/web/repo/issue.go]: Does GitHub also have `May be fixed`?
PR is not merged, so there's no code changes, so why it can be `May be fixed`? There's no fix.

**Comment by yp05327** [comment]
Can we just close the modal after click the button no matter whether the creation is success or not?
Or it is hard to see the error message:
image

**Comment by yp05327** [comment]
image
this is not in middle

**Comment by yp05327** [comment]
image
No word break for long branch name

**Comment by yp05327** [comment]
If I changed the values of the form, and leave this page, a warning will disappear:
image

**Comment by yp05327** [comment]
image
Only branch name here, but no owner name.
ps: `testaaa` branch is from fork.

**Comment by yp05327** [comment]
image
use the time template like this?
image

**Comment by yp05327** [comment]
We have this:
image
So I think we can use it here:
image

File: templates/repo/issue/view_content/sidebar_development.tmpl [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+				<label for="new_branch_name">{{ctx.Locale.Tr "repo.branch.name"}}</label>
+				<input name="new_branch_name" type="text">
+			</div>
+
+			<div class="field">
+				<label for="source_repository">{{ctx.Locale.Tr "repository"}}</label>
+				<div class="ui fluid dropdown selection">
+					<select name="repo_id">
+						<option value=""> </option>
+						{{range .AllowedRepos}}
```
Comment (yp05327) [inline templates/repo/issue/view_content/sidebar_development.tmpl]: If AllowedRepos is empty, then show a create fork button or notification or something else?

File: templates/repo/issue/view_content/sidebar_development.tmpl [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+			<div class="field">
+				<label for="new_branch_name">{{ctx.Locale.Tr "repo.branch.name"}}</label>
+				<input name="new_branch_name" type="text">
+			</div>
+
+			<div class="field">
+				<label for="source_repository">{{ctx.Locale.Tr "repository"}}</label>
+				<div class="ui fluid dropdown selection">
+					<select name="repo_id">
+						<option value=""> </option>
```
Comment (yp05327) [inline templates/repo/issue/view_content/sidebar_development.tmpl]: by default, select the first one?

File: templates/repo/issue/view_content/sidebar_development.tmpl [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+						{{end}}
+					</div>
+				</div>
+			</div>
+
+			<div class="field">
+				<label for="source_branch_name">{{ctx.Locale.Tr "repo.issues.base_branch"}}</label>
+				<div class="ui fluid dropdown selection">
+					<select name="source_branch_name">
+						<option value=""> </option>
```
Comment (yp05327) [inline templates/repo/issue/view_content/sidebar_development.tmpl]: By default, select the first one?

File: templates/repo/issue/view_content/sidebar_development.tmpl (hunk 11 above)
Comment (yp05327) [reply templates/repo/issue/view_content/sidebar_development.tmpl]: As the height will be changed, maybe this should also be fixed.
empty is 38
image
with the value is 42
image

**Comment by yp05327** [comment]
The prefill title and content can be uppercase?

**Comment by yp05327** [comment]
image

Maybe the empty value check can be done by js/browser first?

**Comment by yp05327** [comment]
Branch list is not correct.
image
This branch is only exist in the origin repo, not in the fork repo.
And you can even select the base branch first. It is wrong.

File: templates/repo/issue/view_content/sidebar_development.tmpl [hunk 13]
```diff
@@ -0,0 +1,107 @@
+<span class="text"><strong>{{ctx.Locale.Tr "repo.issues.development"}}</strong></span>
+<div class="ui devlinks list">
+<div>
+<a class="tw-mt-1 fluid ui show-modal" data-modal="#create_branch">{{ctx.Locale.Tr "repo.branch.new_branch"}}</a>
```
Comment (yp05327) [inline templates/repo/issue/view_content/sidebar_development.tmpl]: If the issue is closed, we should not show this.

**Comment by yp05327** [comment]
image
And what is this reference for? It seems that they are similar.

File: templates/repo/issue/view_content/sidebar_development.tmpl (hunk 10 above)
Comment (lunny) [reply templates/repo/issue/view_content/sidebar_development.tmpl]: It will not empty because the original repository has been added to the slice. The suggestion is good but maybe another PR, this PR is big enough.

File: templates/repo/issue/view_content/sidebar_development.tmpl [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+				</div>
+			</div>
+
+			<div class="field">
+				<label for="source_branch_name">{{ctx.Locale.Tr "repo.issues.base_branch"}}</label>
+				<div class="ui fluid dropdown selection">
+					<select name="source_branch_name">
+						{{range .Branches}}
+							<option value="{{.}}"{{if eq . $.Issue.Ref}} checked{{end}}>{{.}}</option>
+						{{end}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/sidebar_development.tmpl]: It is wrong.

1. What is a `Ref` and what was it ever used for? Is it really related to the "issue link"?
2. You can't compare a branch to a ref.

Comment (lunny) [reply templates/repo/issue/view_content/sidebar_development.tmpl]: I have removed it.

**Comment by lunny** [comment]
> The prefill title and content can be uppercase?

This is configured by users' app.ini .

**Comment by lunny** [comment]
I believe all the issues have been addressed. Please review it again. @yp05327 @wxiaoguang

**Comment by yp05327** [comment]
https://github.com/go-gitea/gitea/pull/31899#issuecomment-2412602321
This is not fixed.

**Comment by yp05327** [comment]
Setting:
image
Success message:
image
The branch name is not correct.

ps: there's no `test` branch in `giteaadmin/test22`

**Comment by yp05327** [comment]
How about assign an existing branch? Will do it in another PR?

**Comment by lunny** [comment]
> How about assign an existing branch? Will do it in another PR?

Yes, next step. I think we can also associate `#xxx` with development.

**Comment by yp05327** [comment]
image
The UI seems not good, if there's many PRs or branches, it looks disorganized

**Comment by lunny** [comment]
> [#31899 (comment)](https://github.com/go-gitea/gitea/pull/31899#issuecomment-2412602321)

Wouldn't it be expected?

> [#31899 (comment)](https://github.com/go-gitea/gitea/pull/31899#issuecomment-2412599694) This is not fixed.

Can you post an screenshot? I can't reproduce it. See my screenshots in updated issue content.

**Comment by yp05327** [comment]
> Wouldn't it be expected?

Personally, I don't think it is good.
If you only edited the form and closed the form without creating a new branch, the warning will display.

> Can you post an screenshot? I can't reproduce it. See my screenshots in updated issue content.

image

**Comment by lunny** [comment]
> > Wouldn't it be expected?
> 
> Personally, I don't think it is good. If you only edited the form and closed the form without creating a new branch, the warning will display.

The logic is derived from all other forms. All other forms have the same behaviour to avoid user losing their content accidentally.

> 
> > Can you post an screenshot? I can't reproduce it. See my screenshots in updated issue content.
> 
> image

Looks like it's a common bug which caused almost all the dropdown list.
图片

**Comment by lunny** [comment]
> image The UI seems not good, if there's many PRs or branches, it looks disorganized

Spaces added.

**Comment by yp05327** [comment]
image

image
Locale is missing

**Comment by lunny** [comment]
> image
> 
> image Locale is missing

Fixed in f2a28d0980d1d36b4aebd3d524b4cbcec65f2438

**Comment by yp05327** [comment]
> Fixed in https://github.com/go-gitea/gitea/commit/f2a28d0980d1d36b4aebd3d524b4cbcec65f2438

Not only one. All of them are not added.

**Comment by lunny** [comment]
> > Fixed in [f2a28d0](https://github.com/go-gitea/gitea/commit/f2a28d0980d1d36b4aebd3d524b4cbcec65f2438)
> 
> Not only one. All of them are missing.

These locale strings should be there.
图片

File: models/issues/issue_dev_link.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+	IssueDevLinkTypeBranch IssueDevLinkType = iota + 1
+	IssueDevLinkTypePullRequest
+)
+
+type IssueDevLink struct {
+	ID            int64 `xorm:"pk autoincr"`
+	IssueID       int64 `xorm:"INDEX"`
+	LinkType      IssueDevLinkType
+	LinkedRepoID  int64                  `xorm:"INDEX"` // it can link to self repo or other repo
+	LinkIndex     string                 // branch name, pull request number or commit sha
```
Comment (yp05327) [inline models/issues/issue_dev_link.go]: We have branch table, so this can be int64?
ps: Commit SHA seems meaningless

Comment (lunny) [reply models/issues/issue_dev_link.go]: In future PRs, we can link a commit to this issue. So it will be kept to support that. BTW: Azure supports linking to a commit. I think for some development model, they just push commits with no feature branches or pull requests, they can use this feature.

Comment (yp05327) [reply models/issues/issue_dev_link.go]: Then it can be split into to columns. `LinkIndex`(int64) and `LinkRef`(string)?
Using `string` type means you need to convert it from/to `int64` type every time if it is a PR ID or branch ID which seems not good.

Comment (lunny) [reply models/issues/issue_dev_link.go]: > Then it can be split into to columns. `LinkIndex`(int64) and `LinkRef`(string)? Using `string` type means you need to convert it from/to `int64` type every time if it is a PR ID or branch ID which seems not good.

The conversion is on the memory and we don't know what data we will store according to different LinkType. And this column will never be used as index, so we don't need to consider performance or as search condition.

**Comment by yp05327** [comment]
This section should be hidden when repo is empty.

**Comment by lunny** [comment]
> This section should be hidden or disabled when repo is empty or repo is archive or repo is mirror.

The links are still useable. Only `create branch` link should be hide.

File: models/issues/issue_dev_link.go (hunk 15 above)
Comment (wxiaoguang) [reply models/issues/issue_dev_link.go]: I also agree it doesn't seem to be a good name or proper design. There are already too many "index" in code: search index, database index, issue index, etc.

And using "string" to store "int" in database would cause problems, if you write `("where str_field = ?", intVal)`, then it can't use the index and would cause imperceptible performance problems.

File: models/issues/issue_dev_link.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+	"code.gitea.io/gitea/models/db"
+	git_model "code.gitea.io/gitea/models/git"
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+type IssueDevLinkType int
+
+const (
+	IssueDevLinkTypeBranch IssueDevLinkType = iota + 1
```
Comment (wxiaoguang) [inline models/issues/issue_dev_link.go:19]: Don't understand why it should link to a branch without PR.

`Create a branch from an issue, with pull/merge request #20226` said that "a branch with PR"

Comment (wxiaoguang) [inline models/issues/issue_dev_link.go:18]: Just tried to play with GitHub's development sidebar for a while. I can see there could be a lot of edge cases:

1. create a new branch `123-my-issue`, rename the branch to `123-my-issue-other`, the link disappears, rename another branch to `123-my-issue`, the link won't appear again.
2. create a new branch `123-my-issue`, create a PR from `123-my-issue`, then the link is updated to PR, the branch link is replaced.

I believe these details need enough documents(comments) and tests.

**Comment by wxiaoguang** [comment]
I guess it won't catch 1.23 milestone since this PR doesn't seem to be complete enough.

**Comment by KemptonM** [comment]
Progress on this? We are using Gitea at work and I would love to have this feature

**Comment by lunny** [comment]
I think I need to migrate the current `ref` column in the issue table.

**Comment by krokets** [comment]
Hi, still plan for 1.24.0 ?

**Comment by lunny** [comment]
> Hi, still plan for 1.24.0 ?

Yes.

**Comment by krokets** [comment]
Hi, do you know why this has been postponed to next release?

**Comment by lunny** [comment]
> Hi, do you know why this has been postponed to next release?

This needs more work.

**Comment by lunny** [comment]
I think this is ready to review again.

File: services/pull/pull.go [hunk 17]
```diff
@@ -128,6 +128,35 @@ func NewPullRequest(ctx context.Context, opts *NewPullRequestOptions) error {
 			return err
 		}
 
+		// add dev links
+		if pr.Flow == issues_model.PullRequestFlowGithub {
```
Comment (wxiaoguang) [inline services/pull/pull.go:133]: Why GitHub-only ? Agit PR can't be used to close an issue?

File: services/repository/branch.go [hunk 18]
```diff
@@ -508,6 +508,22 @@ func CanDeleteBranch(ctx context.Context, repo *repo_model.Repository, branchNam
 	return nil
 }
 
+func DeleteIssueDevLinkByBranchName(ctx context.Context, repoID int64, branchName string) error {
+	awBranch, err := git_model.GetBranch(ctx, repoID, branchName)
+	if err != nil && !git_model.IsErrBranchNotExist(err) {
+		return fmt.Errorf("GetBranch: %vc", err)
```
Comment (wxiaoguang) [inline services/repository/branch.go]: What is `%vc`

File: services/repository/branch.go [hunk 19]
```diff
@@ -508,6 +508,22 @@ func CanDeleteBranch(ctx context.Context, repo *repo_model.Repository, branchNam
 	return nil
 }
 
+func DeleteIssueDevLinkByBranchName(ctx context.Context, repoID int64, branchName string) error {
+	awBranch, err := git_model.GetBranch(ctx, repoID, branchName)
```
Comment (wxiaoguang) [inline services/repository/branch.go]: What is `aw` prefix

File: templates/repo/issue/sidebar/development.tmpl [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+	<span class="text"><strong>{{ctx.Locale.Tr "repo.issues.development"}}</strong></span>
+	<div class="ui devlinks list">
+	{{/* AllowedRepos not empty means login user can create new branch in some of the repositories */}}
+	{{if .ShowCreateBranchLink}}
+		<div class="tw-items-center">
+			<a class="tw-mt-1 fluid ui show-modal" data-modal="#create_branch">{{ctx.Locale.Tr "repo.branch.new_branch"}}</a>
+		</div>
+	{{end}}
+	{{range .DevLinks}}
```
Comment (wxiaoguang) [inline templates/repo/issue/sidebar/development.tmpl:12]: Please don't use the data from root template data.

Most side bar parts have been refactored to use their own "struct" types to make code stable.

**Comment by wxiaoguang** [comment]
> I think this is ready to review again.

I don't.

* https://github.com/go-gitea/gitea/pull/31899/files#r1881918670

**Comment by bircni** [comment]
@lunny updates on that?

File: services/pull/pull.go (hunk 17 above)
Comment (lunny) [reply services/pull/pull.go:133]: Agit PR have no real head branch. So that `branch, err := git_model.GetBranch(ctx, pr.HeadRepoID, pr.HeadBranch)` will fail.

File: services/repository/branch.go (hunk 18 above)
Comment (lunny) [reply services/repository/branch.go]: 1a6bfece8d

File: services/repository/branch.go (hunk 19 above)
Comment (lunny) [reply services/repository/branch.go]: 1a6bfece8d

**Comment by lunny** [comment]
> @lunny updates on that?

Updated.

**Review by wxiaoguang — CHANGES_REQUESTED** [review]
https://github.com/go-gitea/gitea/pull/31899/files#r1881918670

> Just tried to play with GitHub's development sidebar for a while. I can see there could be a lot of edge cases:
> 
> 1. create a new branch `123-my-issue`, rename the branch to `123-my-issue-other`, the link disappears, rename another branch to `123-my-issue`, the link won't appear again.
> 2. create a new branch `123-my-issue`, create a PR from `123-my-issue`, then the link is updated to PR, the branch link is replaced.
> 
> I believe these details need enough documents(comments) and tests.

File: services/pull/pull.go (hunk 17 above)
Comment (wxiaoguang) [reply services/pull/pull.go:133]: Is it really reasonable?

Here you are creating "IssueDevLinkTypePullRequest" dev-links.

You mean that AGit PR can't be used to link to an issue and close the related issue?

**Comment by claesrosell** [comment]
I just downloaded this branch to play around with the development sidebar in its current state, to test the edge cases that wxiaoguang mentioned earlier.

The rename and delete branch scenarios behave very differently from the described GitHub behavior. This is not necessarily a bad thing, though:

1. Create a new branch (123-my-issue) and rename it (123-my-issue-other): The link still works, pointing to the same branch with its new name. If you delete the branch, the link disappears. If a new branch with the same name is created, the link will not magically appear again. Personally I think this behavior is better then GitHub

2. Create a new branch (123-my-issue) and create a PR from that branch: The PR shows up in the development sidebar, but the branch (with a "create PR" button) is still there. This branch should probably be hidden when a PR for it is visible in the sidebar.
